File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    
    
    void solve(vector<int>& arr,int idx,int target,vector<vector<int>>& ans,vector<int>& temp){
        
        if(idx==arr.size()){
            if(target==0){
                ans.push_back(temp);
                
            }
            return;
        }
        
        
        // int curr = arr[idx];
        
        if(arr[idx]<=target){ // if curr element is less than target
            
            temp.push_back(arr[idx]); // push in temp
            // target = target-arr[idx];
            solve(arr,idx,target-arr[idx],ans,temp); // decrease target and dont increase idx as we can reuse
            temp.pop_back(); 
            
        }
        // idx+=1;
        solve(arr,idx+1,target,ans,temp);
        
        
        
        
        
    }
    
    
    
    vector<vector<int>> combinationSum(vector<int>& arr, int target) {
        
        vector<vector<int>> ans;
        
        vector<int> temp;
        
        solve(arr,0,target,ans,temp);
        return ans;
    }
};

File: 39-combination-sum/39-combination-sum.py
class Solution:
    
    def solve(self,idx,target,nums,temp,ans):
        
        if(idx==len(nums)):
            if(target==0):
                ans.append(list(temp));
            
            return
        
        if(nums[idx]<=target):
            
            temp.append(nums[idx])
            self.solve(idx,target-nums[idx],nums,temp,ans)
            temp.pop()
        
        self.solve(idx+1,target,nums,temp,ans);
            
            
    
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        idx = 0
        temp = []
        ans = []
        
        self.solve(idx,target,nums,temp,ans);
        return ans
        
        